spanner/metrics.go

```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2024 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spanner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/noop"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"cloud.google.com/go/spanner/internal"
+)
+
+const (
+	builtInMetricsMeterName = "gax-go"
+
+	nativeMetricsPrefix = "spanner.googleapis.com/internal/client/"
+
+	// Monitored resource labels
+	monitoredResLabelKeyProject        = "project_id"
+	monitoredResLabelKeyInstance       = "instance_id"
+	monitoredResLabelKeyInstanceConfig = "instance_config"
+	monitoredResLabelKeyLocation       = "location"
+
+	// Metric labels
+	metricLabelKeyClientUID         = "client_uid"
+	metricLabelKeyClientName        = "client_name"
+	metricLabelKeyDatabase          = "database"
+	metricLabelKeyMethod            = "method"
+	metricLabelKeyStatus            = "status"
+	metricLabelKeyDirectPathEnabled = "directpath_enabled"
+	metricLabelKeyDirectPathUsed    = "directpath_used"
+
+	// Metric names
+	metricNameOperationLatencies = "operation_latencies"
+	metricNameAttemptLatencies   = "attempt_latencies"
+	metricNameOperationCount     = "operation_count"
+	metricNameAttemptCount       = "attempt_count"
+
+	// Metric units
+	metricUnitMS    = "ms"
+	metricUnitCount = "1"
+)
+
+// These are effectively const, but for testing purposes they are mutable
+var (
+	// duration between two metric exports
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName = fmt.Sprintf("go-spanner v%v", internal.Version)
+
+	bucketBounds = []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0, 16.0, 20.0, 25.0, 30.0, 40.0,
```

Comment:
See internal discussion - please sync with Surbhi on the default buckets we use.

Reply:
Took reference from Java PR https://github.com/googleapis/java-spanner/pull/3125/files#diff-be4eb8b7d32f038815df792bedaf4b93ee596806ed2aeaccec8afbcf89b2f342R100-R103

cc: @surbhigarg92